macaroon_recipes.go

```diff
@@ -0,0 +1,112 @@
+package lndclient
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+var (
+	// supportedSubservers is a map of all RPC (sub)server names that are
+	// supported by the lndclient library and their implementing interface
+	// type. We use reflection to look up the methods implemented on those
+	// interfaces to find out which permissions are needed for them.
+	supportedSubservers = map[string]interface{}{
+		"lnrpc":       (*LightningClient)(nil),
+		"chainrpc":    (*ChainNotifierClient)(nil),
+		"invoicesrpc": (*InvoicesClient)(nil),
+		"routerrpc":   (*RouterClient)(nil),
+		"signrpc":     (*SignerClient)(nil),
+		"verrpc":      (*VersionerClient)(nil),
+		"walletrpc":   (*WalletKitClient)(nil),
+	}
+
+	// renames is a map of renamed RPC method names. The key is the name as
+	// implemented in lndclient and the value is the original name of the
+	// RPC method defined in the proto.
+	renames = map[string]string{
+		"ChannelBackup":          "ExportChannelBackup",
+		"ChannelBackups":         "ExportAllChannelBackups",
+		"ConfirmedWalletBalance": "WalletBalance",
+		"Connect":                "ConnectPeer",
+		"DecodePaymentRequest":   "DecodePayReq",
+		"EstimateFeeToP2WSH":     "EstimateFee",
+		"ListTransactions":       "GetTransactions",
+		"PayInvoice":             "SendPaymentSync",
+		"UpdateChanPolicy":       "UpdateChannelPolicy",
+	}
+)
+
+// MacaroonRecipe returns a list of macaroon permissions that is required to use
+// the full feature set of the given list of RPC package names.
+func MacaroonRecipe(c LightningClient, packages []string) ([]MacaroonPermission,
+	error) {
+
+	// Get the full map of RPC URIs and the required permissions from the
+	// backing lnd instance.
+	allPermissions, err := c.ListPermissions(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	uniquePermissions := make(map[string]map[string]struct{})
+	for _, pkg := range packages {
+		// Get the typed pointer from our map of supported interfaces.
+		ifacePtr, ok := supportedSubservers[pkg]
+		if !ok {
+			return nil, fmt.Errorf("unknown subserver %s", pkg)
+		}
+
+		// From the pointer type we can find out the interface, its name
+		// and what methods it declares.
+		ifaceType := reflect.TypeOf(ifacePtr).Elem()
+		serverName := strings.ReplaceAll(ifaceType.Name(), "Client", "")
```

Comment:
I like that we can use reflection to gather method names but maybe a simple constant map of rpc -> method name would also suffice (and is easier to check/debug). Is perfectly fine as is, just a comment.

Reply:
If the list of exceptions keeps growing, I agree that we can switch over to a full map instead. But for now, this keeps the code quite small and also doesn't require you to add/change anything, if you add a new method that's called the same as the RPC it calls.

Reply:
> if you add a new method that's called the same as the RPC it calls

nittynit: can see this being forgotten, maybe add a PR template with a reminder about this? I find that pretty useful. Can be in a follow up.

Reply:
The unit tests should catch that. But to avoid an extra git push cycle I added it to the checklist :+1: